src/libos/src/fs/rootfs.rs

```diff
@@ -242,3 +245,109 @@ fn open_or_create_sefs_according_to(
     };
     Ok(sefs)
 }
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+#[allow(non_camel_case_types)]
+pub struct user_rootfs_config {
+    upper_layer_path: *const i8,
+    lower_layer_path: *const i8,
+    entry_point: *const i8,
+    hostfs_source: *const i8,
+    hostfs_target: *const i8,
+}
+
+impl user_rootfs_config {
+    pub fn from_raw_ptr(ptr: *const user_rootfs_config) -> Result<user_rootfs_config> {
+        let config = unsafe { *ptr };
+        Ok(config)
+    }
+}
+
+fn to_option_pathbuf(path: *const i8) -> Result<Option<PathBuf>> {
+    let path = if path.is_null() {
+        None
+    } else {
+        Some(PathBuf::from(
+            from_user::clone_cstring_safely(path)?
+                .to_string_lossy()
+                .into_owned(),
+        ))
+    };
+
+    Ok(path)
+}
+
+pub fn gen_config_app(config: &user_rootfs_config) -> Result<ConfigApp> {
```

Comment:
I think this implementation is not Object-Oriented, you should implement a `From` for `user_rootfs_config` to transform into `ConfigApp`.  BTW, the structs here should be defined in `config.rs` instead of this file.

Reply:
I will see how to address this.